src/main/java/com/limechain/trie/decoder/TrieHeaderDecoder.java

```diff
@@ -2,43 +2,58 @@
 
+/**
+ * This class is used to decode the header of a Trie node from a ScaleCodecReader input stream.
+ */
+@UtilityClass
 public class TrieHeaderDecoder {
 
     /**
      * Decodes a node header from a ScaleCodecReader input stream.
+     * See <a href="https://spec.polkadot.network/chap-state#defn-node-header">here</a> for more information
      *
      * @param reader the ScaleCodecReader input stream used to read the encoded node data
      * @return The variant, partial key length and mask read from the stream
-     * @throws TrieDecoderException if an error occurs while reading the children bitmap or the storage value.
+     * @throws TrieDecoderException if an error occurs while reading the node type or storage value.
      */
     public static TrieHeaderDecoderResult decodeHeader(ScaleCodecReader reader) {
         try {
             byte currentByte = reader.readByte();
             TrieHeaderDecoderResult header = TrieHeaderDecoder.decodeHeaderByte(currentByte);
-            int partialKeyLengthHeader = header.getPartialKeyLengthHeader();
-            int partialKeyLengthHeaderMask = header.getPartialKeyLengthHeaderMask();
-            byte variantBits = header.getVariantBits();
+            int partialKeyLengthHeader = header.partialKeyLengthHeader();
+            int partialKeyLengthHeaderMask = header.nodeVariant().getPartialKeyLengthHeaderMask();
 
+            // Empty node
+            if (partialKeyLengthHeaderMask == NodeVariant.EMPTY.bits) {
+                return new TrieHeaderDecoderResult(header.nodeVariant(), 0);
+            }
+
+            // Key length is contained in the first byte
             if (partialKeyLengthHeader < partialKeyLengthHeaderMask) {
-                return new TrieHeaderDecoderResult(variantBits, partialKeyLengthHeader, (byte) 0);
+                return new TrieHeaderDecoderResult(header.nodeVariant(), partialKeyLengthHeader);
             }
 
+            // If the value is greater than the maximum possible value the bits can hold (63 for leafs),
+            // then the value of the next 8 bits are added to the length.
+            // This process is repeated until the next value less than < 255
+            // See https://spec.polkadot.network/chap-state#defn-node-header
             while (true) {
                 int nextByte = reader.readUByte();
                 partialKeyLengthHeader += nextByte;
                 if (partialKeyLengthHeader > MAX_PARTIAL_KEY_LENGTH) {
                     throw new TrieDecoderException("Partial key overflow");
                 }
 
-                // Check if current byte is max byte value
+                // Check if current byte is less than max byte value
                 if (nextByte < 255) {
-                    return new TrieHeaderDecoderResult(variantBits, partialKeyLengthHeader, (byte) 0);
+                    return new TrieHeaderDecoderResult(header.nodeVariant(), partialKeyLengthHeader);
                 }
```

Comment:
Isn't is possible for the last byte to be exactly 255? Loop would continue to iterate. Is that expected behaviour? Maybe we can use `hasNext()` somewhere.

Reply:
I think that's fine and the code will execute as expected. 
<img width="854" alt="image" src="https://github.com/LimeChain/Fruzhin/assets/29047760/fc26b9b2-c4d1-4c59-b563-cc30f5be753b">

Reply:
From here: https://spec.polkadot.network/chap-state#defn-node-header